Mathlib/NumberTheory/LSeries/QuadraticNonvanishing.lean

```diff
@@ -0,0 +1,187 @@
+  Function.update (fun s : ℂ ↦ riemannZeta s * LFunction B.χ s) 1 (deriv (LFunction B.χ) 1)
+
+private lemma F_differentiableAt_of_ne (B : BadChar N) {s : ℂ} (hs : s ≠ 1) :
+    DifferentiableAt ℂ B.F s := by
+  apply DifferentiableAt.congr_of_eventuallyEq
+  · exact (differentiableAt_riemannZeta hs).mul <| differentiableAt_LFunction B.χ s (.inl hs)
+  · filter_upwards [eventually_ne_nhds hs] with t ht using Function.update_noteq ht ..
+
+open ArithmeticFunction in
+/-- `B.F` agrees with the L-series of `zetaMul χ` on `1 < s.re`. -/
+private lemma F_eq_LSeries (B : BadChar N) {s : ℂ} (hs : 1 < s.re) :
+    B.F s = LSeries B.χ.zetaMul s := by
+  rw [F, zetaMul, ← coe_mul, LSeries_convolution']
+  · have hs' : s ≠ 1 := fun h ↦ by simp only [h, one_re, lt_self_iff_false] at hs
+    simp only [ne_eq, hs', not_false_eq_true, Function.update_noteq, B.χ.LFunction_eq_LSeries hs]
+    congr 1
+    · simp_rw [← LSeries_zeta_eq_riemannZeta hs, ← natCoe_apply]
+    · exact LSeries_congr s B.χ.apply_eq_toArithmeticFunction_apply
+  -- summability side goals from `LSeries_convolution'`
+  · exact LSeriesSummable_zeta_iff.mpr hs
+  · exact (LSeriesSummable_congr _ fun h ↦ (B.χ.apply_eq_toArithmeticFunction_apply h).symm).mpr <|
+      ZMod.LSeriesSummable_of_one_lt_re B.χ hs
+
+/-- If `χ` is a bad character, then `F` is an entire function. -/
+private lemma F_differentiable (B : BadChar N) : Differentiable ℂ B.F := by
+  intro s
+  rcases ne_or_eq s 1 with hs | rfl
+  · exact B.F_differentiableAt_of_ne hs
+  -- now need to deal with `s = 1`
+  refine (analyticAt_of_differentiable_on_punctured_nhds_of_continuousAt ?_ ?_).differentiableAt
+  · filter_upwards [self_mem_nhdsWithin] with t ht
+    exact B.F_differentiableAt_of_ne ht
+  -- now reduced to showing *continuity* at s = 1
+  let G := Function.update (fun s ↦ (s - 1) * riemannZeta s) 1 1
+  let H := Function.update (fun s ↦ (B.χ.LFunction s - B.χ.LFunction 1) / (s - 1)) 1
+    (deriv B.χ.LFunction 1)
+  have : B.F = G * H := by
+    ext1 t
+    rcases eq_or_ne t 1 with rfl | ht
+    · simp only [F, G, H, Pi.mul_apply, one_mul, Function.update_same]
+    · simp only [F, G, H, Function.update_noteq ht, mul_comm _ (riemannZeta _), B.hχ, sub_zero,
+      Pi.mul_apply, mul_assoc, mul_div_cancel₀ _ (sub_ne_zero.mpr ht)]
+  rw [this]
+  apply ContinuousAt.mul
+  · simpa only [G, continuousAt_update_same] using riemannZeta_residue_one
+  · exact (B.χ.differentiableAt_LFunction 1 (.inr B.χ_ne)).hasDerivAt.continuousAt_div
+
+/-- The trivial zero at `s = -2` of the zeta function gives that `F (-2) = 0`.
+This is used later to obtain a contradction. -/
+private lemma F_neg_two (B : BadChar N) : B.F (-2 : ℝ) = 0 := by
+  have := riemannZeta_neg_two_mul_nat_add_one 0
+  rw [Nat.cast_zero, zero_add, mul_one] at this
+  rw [F, ofReal_neg, ofReal_ofNat, Function.update_noteq (mod_cast (by omega : (-2 : ℤ) ≠ 1)),
+    this, zero_mul]
+
+end BadChar
+
+/-!
+### The main result
+-/
+
+open scoped ComplexOrder in
```

Comment:
```suggestion

```

Reply:
Other than this trivial suggestion it LGTM, sorry I only just got around to looking at this!